File: Day 98/Solution.py
# 493. Reverse Pairs (#Solved)
# Given an integer array nums, return the number of reverse pairs in the array.

# A reverse pair is a pair (i, j) where:

# 0 <= i < j < nums.length and
# nums[i] > 2 * nums[j].

class Solution:
    def reversePairs(self, nums):
        def mergeSort(low, high):
            if low >= high:
                return 0
            
            mid = (low + high) // 2
            invCount = mergeSort(low, mid) + mergeSort(mid + 1, high)
            
            # Count reverse pairs
            invCount += reverse_pairs(low, mid, high)
            
            # Merge
            merge(low, mid, high)
            
            return invCount
        
        def merge(low, mid, high):
            temp = []
            left, right = low, mid + 1
            
            while left <= mid and right <= high:
                if nums[left] <= nums[right]:
                    temp.append(nums[left])
                    left += 1
                else:
                    temp.append(nums[right])
                    right += 1
            
            temp.extend(nums[left:mid+1])
            temp.extend(nums[right:high+1])
            
            nums[low:high+1] = temp
        
        def reverse_pairs(low, mid, high):
            right = mid + 1
            cnt = 0
            for i in range(low, mid + 1):
                while right <= high and nums[i] > 2 * nums[right]:
                    right += 1
                cnt += (right - (mid + 1))
            return cnt
        
        return mergeSort(0, len(nums) - 1)

if __name__ == "__main__":
    sol = Solution()
    print(sol.reversePairs([1,3,2,3,1]))  # 2
    print(sol.reversePairs([2,4,3]))      # 0

File: Day 98/Solution.java
// 493. Reverse Pairs (#Solved)
// Given an integer array nums, return the number of reverse pairs in the array.

// A reverse pair is a pair (i, j) where:

// 0 <= i < j < nums.length and
// nums[i] > 2 * nums[j].

import java.util.ArrayList;
import java.util.Arrays;

public class Solution {
    public int reversePairs(int[] nums) {
        return mergeSort(nums, 0, nums.length - 1);
    }
    
    private static int mergeSort(int[] arr, int low, int high) {
        if (low >= high) return 0;
        
        int mid = low + (high - low) / 2;
        int invCount = mergeSort(arr, low, mid) + 
                      mergeSort(arr, mid + 1, high);
        
        // Count reverse pairs during merge
        invCount += reversePairs(arr, low, mid, high);
        
        // Actual merge
        merge(arr, low, mid, high);
        
        return invCount;
    }
    
    private static void merge(int[] arr, int low, int mid, int high) {
        ArrayList<Integer> temp = new ArrayList<>();
        int left = low;
        int right = mid + 1;
        
        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                temp.add(arr[left++]);
            } else {
                temp.add(arr[right++]);
            }
        }
        
        while (left <= mid) temp.add(arr[left++]);
        while (right <= high) temp.add(arr[right++]);
        
        for (int i = low; i <= high; i++) {
            arr[i] = temp.get(i - low);
        }
    }
    
    public static int reversePairs(int[] arr, int low, int mid, int high) {
        int right = mid + 1;
        int cnt = 0;
        for (int i = low; i <= mid; i++) {
            while (right <= high && (long)arr[i] > 2L * arr[right]) {
                right++;
            }
            cnt += (right - (mid + 1));
        }
        return cnt;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.reversePairs(new int[]{1,3,2,3,1}));  // 2
        System.out.println(sol.reversePairs(new int[]{2,4,3}));      // 0
    }
}
